core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -230,6 +230,14 @@ public DoubleCacheStream flatMapToDouble(Function<? super R, ? extends DoubleStr
 
    @Override
    public CacheStream<R> limit(long maxSize) {
+      if (maxSize <= 0) {
+         throw new IllegalArgumentException("Limit must be greater than 0, was " + maxSize);
+      }
+      // We can limit the distributed batch size if we know we have 1 value per entry
```

Comment:
Initially I thought this is ok, but then I realized that the distributedBatchSize is not applied as-is to remote nodes, it's divided by (up to) 4. So if the stream has a filter that only matches entries on one remote node, it could take 4 publisher requests to fetch up to limit entries instead of just 1. Is that ok?

Maybe this logic should be at a lower level, e.g. if there is a limit set we could start the publisher only on the local node first, and continue with remote nodes if there aren't enough entries on the local nodes.

Reply:
Agreed. I will remove this commit as it needs to bake more.

Reply:
Unfortunately, the new Publisher stuff provides limited view to when a limit is supplied so the idea really only works with DistributedCacheStream. However, I think I need to take a stab at this later after changing the internals to use Publisher more than streams.